kubelet: add setting for configuring cpuManagerPolicy

gthao313 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Part of #1628

**Description of changes:**
Fluent-bit does not work with `cpuPolicyManager: static`, and it does work with `cpuPolicyManager: none`. We expose `kubernetes.cpu-manager-policy` as a setting, so customers can specify `cpuManagerPolicy` to `none` and not be blocked by this issue while running fluent bit.

- [x] expose `kubernetes.cpu-manager-policy` as a setting
- [x] expose `kubernetes.cpu-manager-reconcile-policy` as a setting
- [x] add migration for cpu-manager-manager-policy - deal with the issue that changing the CPU manager policy caused issues on upgrade/downgrade ( from none to static and from static to none)
- [x] add migration for cpu-manager-reconcile-period

#### Solution about deal with the issue that changing the CPU manager policy caused issues on upgrade/downgrade
error
```
2028 cpu_manager.go:203] [cpumanager] could not initialize checkpoint manager: could not restore state from checkpoint: configured policy "static" differs from state checkpoint policy "none", please drain this node and delete the CPU manager checkpoint file "/var/lib/kubelet/cpu_manager_state" before restarting Kubelet, please drain node and remove policy state file
```

The reason why upgrade/downgrade of CPU manager policy migration is configured policy "static" differs from state checkpoint policy "none". Therefore, I use migration tool to delete cpu-manager-policy checkpoint for every migration process. No orchestrated workload would be running when the migrator is running, so it should be safe to delete that file before kubelet starts.

**Testing done:**

- [x] Test behavioral effect of new setting `cpu-manager-policy`
- [x] Test behavioral effect of new setting `cpu-manager-reconcile-policy`
- [x] Test migration process - `cpu-manager-policy`
- [x] Test migration process - `cpu-manager-reconcile-policy`

#### `cpu-manager-policy`
launching instance with the AMI which contains new setting and check if `cpu-manager-policy` has been configured expectantly. Meanwhile,  configure it as `none` and test if fluent-bit pod work as expected.

Step1 check if `cpu-manager-policy` has been configured expectantly:

Test with `cpu-manager-policy= "none"`
```
1726 policy_none.go:43] [cpumanager] none policy: Start
```
Test with default ("none")
```
1737 cpu_manager.go:193] [cpumanager] starting with none policy
```
Test with `cpu-manager-policy= "static"`
```
1723 cpu_manager.go:193] [cpumanager] starting with static policy
```

Step2 configure it as `none` and test if fluent-bit pod work as expected:
1. kubectl run --image public.ecr.aws/aws-observability/aws-for-fluent-bit:2.16.0 --command -- sh
2. kubectl exec -ti sh -- sh
3. /fluent-bit/bin/fluent-bit -H -i cpu -o null 
4. while curl localhost:2020; do echo pass; done

result (web server works perfectly ):
```
COMMAND   PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
fluent-bi   1 root   19u  IPv4  20229      0t0  TCP *:24224 (LISTEN)
fluent-bi  36 root   30u  IPv4  19367      0t0  TCP *:2020 (LISTEN)
```

#### Migration test:
upgrade
Step1: Upgrade to v1.1.3
```
bash-5.0# updog check-update -a --json
[
  {
    "variant": "aws-k8s-1.20",
    "arch": "x86_64",
    "version": "1.1.3",
    "max_version": "1.1.3",
.....

```
cat var/lib/kubelet/cpu_manager_state
{"policyName":"static","defaultCpuSet":"0-3","checksum":611748604}

```
bash-5.0# updog update -i 1.1.3 -r -n
Starting update to 1.1.3
```

Step2: check if cpu-manager-policy works as expected after upgrading
```
cat var/lib/kubelet/cpu_manager_state
{"policyName":"none","defaultCpuSet":"","checksum":1353318690}
```

```
journalctl - u kubelet
1701 policy_none.go:43] [cpumanager] none policy: Start
```

Step3: Specify new setting `cpu-manage-policy`through control container
```
apiclient set -j '{"kubernetes": {"cpu-manager-policy": "none"}}'
```
```
bash-5.0# cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy
"none"
```

downgrade

Step1: Check migration binary
```
ls -al /var/lib/bottlerocket-migrations
total 492
drwx------.  2 root root   4096 Jun 29 21:51 .
drwxr-xr-x. 14 root root   4096 Jun 29 22:22 ..
-rw-r--r--.  1 root root 488744 Jun 29 22:21 0b168f448fed442fc3be61059256d4c12a423aa6afba5ec825596522dd015d41.migrate_v1.1.3_cpu-manager-policy.lz4

```

Step2: Downgrade to previous verison
```
signpost rollback-to-inactive
reboot
```

Step3: check if cpu-manager-policy works as expected after downgrading
```
cat var/lib/kubelet/cpu_manager_state
{"policyName":"static","defaultCpuSet":"0-3","checksum":611748604}
```

```
journalctl - u kubelet
1700 cpu_manager.go:193] [cpumanager] starting with static policy
```

Step4: Check if `cpu-manager-policy` have been removed
```
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy
cat: /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy: No such file or directory
```

#### `cpu-manager-reconcile-period`
launching instance with the AMI which contains new setting and check if `cpu-manager-reconcile-period` has been configured expectantly.

Step1 check if `cpu-manager-reconcile-period` has been configured expectantly:

Test with `cpu-manager-reconcile-period = 100s`
```
cpu_manager.go:194] [cpumanager] reconciling every 100s
```
Test with `cpu-manager-reconcile-period = 10m`
```
1730 cpu_manager.go:194] [cpumanager] reconciling every 10m0s
```
Test with `cpu-manager-reconcile-period = 1h`
```
1736 cpu_manager.go:194] [cpumanager] reconciling every 1h0m0s
```
Test with `cpu-manager-reconcile-period = 1h0m20s`
```
cpu_manager.go:194] [cpumanager] reconciling every 1h0m20s
```
Test with `cpu-manager-reconcile-period = 10m20s`
```
1726 cpu_manager.go:194] [cpumanager] reconciling every 10m20s
```
Test with `cpu-manager-reconcile-period = 1h10m2s`
```
1727 cpu_manager.go:194] [cpumanager] reconciling every 1h10m2s
```
#### Migration test:
upgrade
Step1: Upgrade to v1.1.3
```
bash-5.0# updog check-update -a --json
[
  {
    "variant": "aws-k8s-1.20",
    "arch": "x86_64",
    "version": "1.1.3",
    "max_version": "1.1.3",
.....

bash-5.0# updog update -i 1.1.3 -r -n
Starting update to 1.1.3
```
Step2: Specify new setting `cpu-manager-reconcile-period`through control container
```
apiclient set -j '{"kubernetes": {"cpu-manager-reconcile-period": "10m"}}'
apiclient set -j '{"kubernetes": {"cpu-manager-policy": "none"}}'
```

```
bash-5.0# cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-reconcile-period
"10m"
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy
"none"
```

```
journalctl -u kubelet
cpu_manager.go:194] [cpumanager] start with none policy
cpu_manager.go:194] [cpumanager] reconciling every 10m0s
```

downgrade
Step1: Check migration binary
```
ls -al /var/lib/bottlerocket-migrations
0b168f448fed442fc3be61059256d4c12a423aa6afba5ec825596522dd015d41.migrate_v1.1.3_cpu-manager-reconcile-period.lz4
0b168f448fed442fc3be61059256d4c12a423aa6afba5ec825596522dd015d41.migrate_v1.1.3_cpu-manager-policy.lz4

```
Step2: Downgrade to previous verison
```
signpost rollback-to-inactive
reboot
```

Step3: Check if `cpu-manager-policy` have been removed
```
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-reconcile-period
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-reconcile-period: No such file or directory
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy
cat /var/lib/bottlerocket/datastore/current/live/settings/kubernetes/cpu-manager-policy: No such file or directory
```

```
journalctl -u kubelet
cpu_manager.go:194] [cpumanager] start with static policy
1702 cpu_manager.go:194] [cpumanager] reconciling every 10s
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

gthao313 committed:
kubelet: add setting for configuring cpuManagerPolicy

pass cpu-manager-policy argument to kubelet

gthao313 committed:
fix cpu-manager-policy unusual checkpoint

upgrade: remove the cpu manager policy state file on boot
downgrade: add a migration `kubelet-cpu-manager-state` to remove the
cpu manager policy state file

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -74,6 +74,12 @@ pub mod error {
         },
+
+        #[snafu(display("Invalid Cpu Manager policy '{}'", input))]
+        InvalideCpuManagerPolicy {
```
```suggestion
        InvalidCpuManagerPolicy {
```

bcressey commented on README.md:
```diff
@@ -358,6 +358,7 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.kube-api-burst`: The burst to allow while talking with kubernetes.
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `static`.
```
I'd like to default to "none" here given known incompatibilities with fluent-bit as it's deployed today. We can always change the default for future Kubernetes variants.

bcressey commented on packages/kubernetes-1.20/kubelet-config:
```diff
@@ -74,7 +74,7 @@ systemReserved:
   {{~/each}}
 {{~/if}}
-cpuManagerPolicy: "static"
+cpuManagerPolicy: {{default "static" settings.kubernetes.cpu-manager-policy}}
```
When we originally rolled this out, we reverted it because changing the CPU manager policy caused issues on upgrade.

There's some local state that will need to be fixed up either as part of the migration, or as a helper command to check for a difference between the on-disk value and the configured value. The latter might make this setting possible to change on a running system, if still not entirely safe.

bcressey reviewed (commented):
I'd like to add a setting for `cpu-manager-reconcile-period` while we're at it.

samjo-nyang commented:
Could you backport this option to k8s 1.18 and 1.19? I think the migration process will not make any problems if you set "none" as default value.

gthao313 committed:
kubelet: add setting for configuring cpuManagerReconcilePeriod

pass cpu-manager-reconcile-period argument to kubelet

gthao313 committed:
migrations: add migrations for cpu-manager-policy and cpu-manager-reconcile-period

adds migration for two new settings `settings.kubernetes.cpu-manager-policy`
and `settings.kubernetes.cpu-manager-reconcile-period`

gthao313 commented:
Push above 

- [x] add a new setting `cpu-manager-reconcile-period` and it migration.
- [x] add a custom migration for `cpu-manger-policy` to deal with the issue that changing the CPU manager policy caused issues on upgrade/downgrade
- [x] Default  `cpu-manger-policy` to `none`.
- [x] Support `cpu-manger-policy` on 1.16, 1.17, 1.18 and 1.19 variant.

`cpu-manger-policy` migration logic: The reason why upgrade/downgrade of CPU manager policy migration is configured policy "static" differs from state checkpoint policy "none". Therefore, I use migration tool to delete `cpu-manager-policy checkpoint` during migration process. No orchestrated workload would be running when the migrator is running, so it should be safe to delete that file before kubelet starts.
```
2028 cpu_manager.go:203] [cpumanager] could not initialize checkpoint manager: could not restore state from checkpoint: configured policy "static" differs from state checkpoint policy "none", please drain this node and delete the CPU manager checkpoint file "/var/lib/kubelet/cpu_manager_state" before restarting Kubelet, please drain node and remove policy state file

```

webern commented on packages/kubernetes-1.20/kubelet-config:
```diff
@@ -74,7 +74,10 @@ systemReserved:
   {{~/each}}
 {{~/if}}
-cpuManagerPolicy: "static"
+cpuManagerPolicy: {{default "none" settings.kubernetes.cpu-manager-policy}}
```
Are we intentionally changing the default to "none"? I thought we would expose the setting but otherwise leave the default unchanged?

```suggestion
cpuManagerPolicy: {{default "static" settings.kubernetes.cpu-manager-policy}}
```

gthao313 replied:
We currently default it to `none` here given known incompatibilities with fluent-bit as it's deployed today. We can change the default  to `static` in future once fluent-bit bug has been fixed.

webern replied:
We heard from @bcressey out-of-band that he _does_ want to change the default back to `none` so I am resolving this issue and approving.

webern commented on packages/kubernetes-1.21/kubelet-config:
```diff
@@ -74,7 +74,10 @@ systemReserved:
   {{~/each}}
 {{~/if}}
-cpuManagerPolicy: "static"
+cpuManagerPolicy: {{default "none" settings.kubernetes.cpu-manager-policy}}
```
Same question as above.

```suggestion
cpuManagerPolicy: {{default "static" settings.kubernetes.cpu-manager-policy}}
```

webern replied:
We heard from @bcressey out-of-band that he _does_ want to change the default back to `none` so I am resolving this issue and approving.

webern commented on sources/api/migration/migrations/v1.1.3/cpu-manager-policy/src/main.rs:
```diff
@@ -0,0 +1,65 @@
+                "Deleting existing cpu manager policy checkpoint: '{}'",
+                CPU_MANAGER_POLICY_CHECKPOINT
+            );
+            fs::remove_file(CPU_MANAGER_POLICY_CHECKPOINT).context(error::RemoveFile {
+                path: CPU_MANAGER_POLICY_CHECKPOINT,
+            })?;
```
I don't think the program should exit with error if the file does not exist. I think the right way to do this is to check the std::io::ErrorKind for NotFound and maybe println! if the ErrorKind is NotFound, otherwise error out. (Same for the `backward` function below as well)

https://doc.rust-lang.org/std/fs/fn.remove_file.html
https://doc.rust-lang.org/std/io/enum.ErrorKind.html

gthao313 replied:
Sure. I'll update that! thanks

gthao313 commented:
Push above improving error handle for `fileremove` according to Matt's idea.

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// CpuManagerReconcilePeriod represents a string that contains a valid cpu management reconcile period.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct CpuManagerReconcilePeriod {
+    inner: String,
+}
+
+lazy_static! {
+    pub(crate) static ref CPU_MANAGER_RECONCILE_PERIOD: Regex = Regex::new(
+        r"^(([0-9.]+)(h))?(([0-9.]+)(m))?(([0-9.]+)(s))?$"
+    ).unwrap();
+}
```
I'd call this type `KubernetesDurationValue` since there are several other settings that could use this same validation logic.

bcressey commented on sources/api/migration/migrations/v1.1.3/cpu-manager-policy/src/main.rs:
```diff
@@ -0,0 +1,84 @@
+use std::process;
+
+const SETTING: &str = "settings.kubernetes.cpu-manager-policy";
+const CPU_MANAGER_POLICY_CHECKPOINT: &str = "var/lib/kubelet/cpu_manager_state";
```
```suggestion
const CPU_MANAGER_POLICY_CHECKPOINT: &str = "/var/lib/kubelet/cpu_manager_state";
```

bcressey commented on sources/api/migration/migrations/v1.1.3/cpu-manager-policy/src/main.rs:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    fn backward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
+        // removing cpu-manager-policy setting, which old version doesn't know about it.
+        if let Some(data) = input.data.remove(SETTING) {
+            println!("Removed {}, which was set to '{}'", SETTING, data);
+        } else {
+            println!("Found no {} to remove", SETTING);
+        }
```
I'd prefer to handle removing the setting together with `cpu-manager-reconcile-period`, and leave this migration purely for the unusual checkpoint fix.

Maybe name them:
* kubelet-cpu-manager
* kubelet-cpu-manager-state

bcressey commented on sources/api/migration/migrations/v1.1.3/cpu-manager-policy/src/main.rs:
```diff
@@ -0,0 +1,84 @@
+
+const SETTING: &str = "settings.kubernetes.cpu-manager-policy";
+const CPU_MANAGER_POLICY_CHECKPOINT: &str = "var/lib/kubelet/cpu_manager_state";
+/// We remove cpu manager policy checkpoint value if checkpoint value is differenet from upgrade version
```
```suggestion
/// We remove cpu manager policy checkpoint value if checkpoint value is different from upgrade version
```

bcressey reviewed (commented):
For the CPU state migration, it might be better to just always remove the state file on boot, when we're sure no pods are running.

I'd recommend trying it without the migration, and adding this line to all the `kubernetes-tmpfiles.conf` snippets instead:
```
r! /var/lib/kubelet/cpu_manager_state
```

That still doesn't make the setting safe to switch on a running system, but it at least means a reboot will always fix it.

bcressey commented:
> I'd recommend trying it without the migration, and adding this line to all the `kubernetes-tmpfiles.conf` snippets instead:

Oops - we'll still need the migration to fix up the downgrade path when going back to a version that didn't have the `tmpfiles` line, so we should do both.

gthao313 commented:
Push above 

- [x] rename variables like  `KubernetesDurationValue `
- [x] improve `cpu-manager-policy` unusual checkpoint to  **For upgrade, add r! /var/lib/kuberlet/cpu_manager_state  and for downgrade we should still use migration backward**
- [x] typo improvement

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -113,12 +113,11 @@ use std::collections::HashMap;
-    KubernetesLabelValue, KubernetesQuantityValue, KubernetesReservedResourceKey,
-    KubernetesTaintValue, KubernetesThresholdValue, Lockdown, SingleLineString, SysctlKey, Url,
-    ValidBase64,
+    CpuManagerPolicy, BootstrapContainerMode, DNSDomain, ECSAgentLogLevel,
```
nit on ordering:
```suggestion
    BootstrapContainerMode, CpuManagerPolicy, DNSDomain, ECSAgentLogLevel,
```

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+}
+
+lazy_static! {
+    pub(crate) static ref CPU_MANAGER_RECONCILE_PERIOD: Regex = Regex::new(
```
```suggestion
    pub(crate) static ref KUBERNETES_DURATION_VALUE: Regex = Regex::new(
```

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// KubernetesDurationValue represents a string that contains a valid kubernetes duration value.
```
```suggestion
/// KubernetesDurationValue represents a string that contains a valid Kubernetes duration value.
```

etungsten commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+
+const SETTING: &str = "settings.kubernetes.cpu-manager-policy";
+const CPU_MANAGER_POLICY_CHECKPOINT: &str = "/var/lib/kubelet/cpu_manager_state";
+/// We remove cpu manager policy checkpoint value if checkpoint value is different from upgrade version
```
I think we're removing this checkpoint file always. Not only when it's different.

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+
+lazy_static! {
+    pub(crate) static ref CPU_MANAGER_RECONCILE_PERIOD: Regex = Regex::new(
+        r"^(([0-9.]+)(h))?(([0-9.]+)(m))?(([0-9.]+)(s))?$"
```
Do we need all the inner capture groups? We should also support `ms`. Not sure if we should allow cases like "....3h" either. Does kubelet have a regex we could use instead?

```suggestion
        r"^([0-9.]+h)?([0-9.]+m)?([0-9.]+s)?([0-9.]+ms)?$"
```

gthao313 replied:
I don't think kubelet have a regex that we can use since I did not find anything on kubelet source code. I dot not think we should allow cases like `10s10h` in bottlerocket. Although `10s10h` is accepted in kubelet and it will be converted to `10h10s`, `10s10h` is not a meaningful format for me and we should keep a meaningful input standard .

etungsten commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+
+impl Migration for CpuManagerPolicyCleaner {
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        println!("CpuManagerPolicyCleaner forward has no work to do on upgrade.");
```
I think it's best to explain why it doesn't have work to do on upgrade. We're not changing the value of a setting or anything. It's just that we're adding a behavior where we always remove the `cpu_manager_state` checkpoint file, therefore we don't need to explicitly remove the file during forward migration.

bcressey replied:
```suggestion
        println!("CpuManagerPolicyCleaner has no work to do on upgrade.");
```

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -74,6 +74,15 @@ pub mod error {
+            source: serde_plain::Error,
+        },
+
+        #[snafu(display("Invalid Cpu Manager reconcile period '{}'", input))]
```
We might want to reuse `InvalidPlainValue` for this case - we don't want to hardcode "Cpu Manager" here.

bcressey commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+
+const SETTING: &str = "settings.kubernetes.cpu-manager-policy";
+const CPU_MANAGER_POLICY_CHECKPOINT: &str = "/var/lib/kubelet/cpu_manager_state";
+/// We remove cpu manager policy checkpoint value if checkpoint value is different from upgrade version
+/// or downgrade version cpu-manager-policy setting value.
```
```suggestion
/// We remove cpu manager policy checkpoint value on downgrade, since older versions did not
/// clean up this state file on boot.
```

bcressey commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+use std::path::Path;
+use std::process;
+
+const SETTING: &str = "settings.kubernetes.cpu-manager-policy";
```
not used

bcressey commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+        Ok(input)
+    }
+}
+/// We added a new settings for configuring kubelet, `settings.kubernetes.cpu-manager-policy`
```
```suggestion
/// We changed the default for CPU manager policy and need to handle kubelet's state file. 
```

tjkirch commented on README.md:
```diff
@@ -358,6 +358,8 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.kube-api-burst`: The burst to allow while talking with kubernetes.
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`.
```
We should probably document that Kubernetes strongly recommends a full reboot if you change this setting after launch.  Our kubernetes service and kubelet-config configuration-file will notice the update and be rewritten, but kubelet isn't designed to handle that change while the system is running, and things could start to go wrong when it notices the state file's policy doesn't match the config file's policy.

tjkirch commented on README.md:
```diff
@@ -358,6 +358,8 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`.
+* `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period.
```
I think it'd help to give a brief explanation of why the user would care about changing these settings; they're pretty opaque, even in the k8s documentation.

gthao313 replied:
That sounds good. Thanks

tjkirch commented on sources/api/migration/migration-helpers/src/error.rs:
```diff
@@ -99,6 +99,12 @@ pub enum Error {
+
+    #[snafu(display("Failed to delete file '{}': '{}'", path, source))]
+    RemoveFile {
+        path: String,
```
```suggestion
    #[snafu(display("Failed to delete file '{}': {}", path.display(), source))]
    RemoveFile {
        path: PathBuf,
```

gthao313 commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -817,3 +817,53 @@ mod test_cpu_manager_policy {
+
+lazy_static! {
+    pub(crate) static ref KUBERNETES_DURATION_VALUE: Regex = Regex::new(
+        r"^([0-9.]+h)?([0-9.]+m)?([0-9.]+s)?([0-9.]+ms)?$"
```
New changes:

1. simplify regex like wiping off `( )` 
2. support `ms`
2.  set standard that we do not accept formats like  `10s10h`, `10s1m` ....

tjkirch replied:
I believe this would still allow empty input; we should probably have a `ensure!(!foo.is_empty(), error::bla)` check in the `try_from` implementation.

tjkirch replied:
As Ben mentioned, it would also allow values like `...3ms` which we wouldn't want.  I think we can handle that by fixing the decimal checks, maybe something like `r"^(([0-9]+\.)?[0-9]+h)?(([0-9]+\.)?[0-9]+m)?(([0-9]+\.)?[0-9]+s)?(([0-9]+\.)?[0-9]+ms)?$"`

gthao313 reviewed (commented):
Push above improve `KubernetesDurationValue` regex and improve comments/variable name

tjkirch commented on packages/kubernetes-1.16/kubelet-config:
```diff
@@ -74,6 +74,10 @@ systemReserved:
   {{@key}}: "{{this}}"
   {{~/each}}
 {{~/if}}
+cpuManagerPolicy: {{default "none" settings.kubernetes.cpu-manager-policy}}
```
Would you please confirm how you tested this on 1.16 - 1.19?  We weren't listing the setting before, so there's potential for kubelet to behave differently now that we're giving it an explicit value, even if we think it's the same.

gthao313 replied:
Yeah, sure thing.

gthao313 replied:
### Test for 1.16 -1.19 variants
 check if `cpu-manager-policy` has been configured expectantly and test if fluent-bit pod work as expected:

Test with `cpu-manager-policy= "none"`
```
1726 policy_none.go:43] [cpumanager] none policy: Start
```
Test with default ("none")
```
1737 cpu_manager.go:193] [cpumanager] starting with none policy
```
Test with `cpu-manager-policy= "static"`
```
1723 cpu_manager.go:193] [cpumanager] starting with static policy
```

configure it as `none` and test if fluent-bit pod work as expected:
1. kubectl run --image public.ecr.aws/aws-observability/aws-for-fluent-bit:2.16.0 --command -- sh
2. kubectl exec -ti sh -- sh
3. /fluent-bit/bin/fluent-bit -H -i cpu -o null 
4. while curl localhost:2020; do echo pass; done

Migration test - check if `cpu-manager-policy` still be started after upgrade/downgrade even unusual checkpoint value (same as test for 1.20).

tjkirch replied:
We also confirmed offline that regardless of whether you have an explicit "none" setting or the default, you do see both messages, `none policy: Start` and `starting with none policy` - I wanted to make sure we weren't taking a different code path.

tjkirch commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+
+    #[test]
+    fn bad_names() {
+        for err in &["100", "ten second", "1m2h", "9ns", &"a".repeat(23)] {
```
We can also check empty input, and cases like `...3ms` and `1..5s`.

tjkirch commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -768,3 +768,102 @@ mod test_kubernetes_cloud_provider {
+
+    #[test]
+    fn good_tokens() {
+        for ok in &["9ms", "99s", "20m", "1h", "1h2m3s10ms", "4m5s10ms", "2h3s10ms"] {
```
We can also check decimals like `1.5h3.5m` since they'd be allowed.

tjkirch commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,61 @@
+
+    fn backward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        // removing existing cpu_manager_policy_state file
+        if Path::new(CPU_MANAGER_POLICY_CHECKPOINT).exists() {
```
I don't think this check is helping because we already check for NotFound and can indicate that it didn't exist.  I think it'd be better to remove this because then we'd get output from the migration rather than silence.

tjkirch commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,61 @@
+                "Deleting existing cpu manager policy checkpoint: '{}'",
+                CPU_MANAGER_POLICY_CHECKPOINT
+            );
+            match fs::remove_file(CPU_MANAGER_POLICY_CHECKPOINT) {
+                Ok(()) => {}
+                Err(e) => {
```
```suggestion
            if let Err(e) = fs::remove_file(CPU_MANAGER_POLICY_CHECKPOINT) {
```

tjkirch commented on sources/api/migration/migrations/v1.1.3/kubelet-cpu-manager-state/src/main.rs:
```diff
@@ -0,0 +1,61 @@
+                            path: CPU_MANAGER_POLICY_CHECKPOINT,
+                        });
+                    } else {
+                        println! ("NotFound: '{}'", CPU_MANAGER_POLICY_CHECKPOINT)
```
Would you please run `cargo fmt` and `cargo clippy` and confirm it's clean?

gthao313 commented:
Push above:

1. Improve code according to Tom's comments
2. Improve Kubernetes duration regex and unit test.
3. Done configuring setting tests and migration test for kubernetes 1.16 to 1.19 variant. All tests have been passed.

bcressey commented on README.md:
```diff
@@ -358,6 +358,11 @@ The following settings are optional and allow you to further configure your clus
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow
+pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node, you can set this setting to `static`. We recommend
+a full reboot if you change this setting after launch.
+* `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period, which can configure CPU Manager Reconcile Kubelet CPU memory
```
```suggestion
* `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period, which controls how often updated CPU assignments are written to cgroupfs. 
```

webern commented on README.md:
```diff
@@ -358,6 +358,11 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.kube-api-burst`: The burst to allow while talking with kubernetes.
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow
```
nit: We usually wrap at the sentence break. This seems like kind of a wide wrapping.

gthao313 commented:
push above improve README.

tjkirch commented on README.md:
```diff
@@ -358,6 +358,8 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.kube-api-burst`: The burst to allow while talking with kubernetes.
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node, you can set this setting to `static`. We recommend a full reboot if you change this setting after launch.
```
```suggestion
* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node, you can set this setting to `static`. You should reboot if you change this setting after startup - try `apiclient reboot`.
```

tjkirch commented on README.md:
```diff
@@ -358,6 +358,8 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.container-log-max-size`: The maximum size of container log file before it is rotated.
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
+* `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node, you can set this setting to `static`. We recommend a full reboot if you change this setting after launch.
+* `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period, which controls how often updated CPU assignments are written to cgroupfs.
```
```suggestion
* `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period, which controls how often updated CPU assignments are written to cgroupfs.  The value is a duration like `30s` for 30 seconds or `1h5m` for 1 hour and 5 minutes.
```

gthao313 commented:
Push above improve README according to Tom's comments
